### ISSUE TITLE
Add recent locations on the Facilities page 

### ISSUE THREAD

**Issue by ldelacosta** [department-of-veterans-affairs/va.gov-team#80213]
## Description
We would like to improve the user experience when the Veteran is in our scheduling workflow. The facilities page will be modified. We are requesting a new category to select `By recent locations`. Adding the new dropdown will allow the Veteran to view their most recent site(s) visted and hope it brings a positive experience for the user. 

---

## Acceptance Criteria
Background: the `va_online_scheduling_recent_locations_filter` toggle is on

- [ ] Given the user is in the `Start scheduling` workflow
      When the user lands on the `Choose your VA location` page
      Then the user will see the default view `By recent locations`
      And the distance to the facility does not display
      And page design matches the [design spec](https://www.figma.com/file/n3kzz08Icg3zcs4F8JPtIk/Add-Recent-locations-%7C-Appointments-FE?type=design&node-id=1-52&mode=design&t=roInzEUxAwT25kvA-0)

- [ ] Given the user is in the `Start scheduling` workflow
       When the user lands on the `Choose your VA location` page and does not have any recent location
        Then the `By recent locations` option does not display in the list
        And the list is sorted following the current logic - see the [feature reference](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/products/health-care/appointments/va-online-scheduling/feature-reference/schedule-flow/va-direct--choose-location.md#:~:text=Display%20and%20sort%20multiple%20facilities).
        And page design matches the [design spec](https://www.figma.com/file/n3kzz08Icg3zcs4F8JPtIk/Add-Recent-locations-%7C-Appointments-FE?type=design&node-id=1-52&mode=design&t=roInzEUxAwT25kvA-0)

### Technical Details
- vets-api work [#83352](https://github.com/department-of-veterans-affairs/va.gov-team/issues/83352) 
- Check the Upcoming appointments (included canceled booked appointments) for facility information 
- Check Past appointments to retrieve the facility information looking at the user's 12 months history. 
- vets-api will collect the data beforehand and frontend will display recent locations on the locations page
- If no locations returned from vets-api, then frontend will display the `By home address` view.  
- Track the user switching between toggles `By recent location`, `By home address `or `By location`. 

### Design Assets
[Figma file](https://www.figma.com/file/n3kzz08Icg3zcs4F8JPtIk/Add-Recent-locations-%7C-Appointments-FE?type=design&node-id=1-52&mode=design&t=roInzEUxAwT25kvA-0)

## Definition of Done
- [x] All acceptance criteria are met
- [ ] Documentation is updated

---

**Comment by cferris32** [comment]
Related tickets:

https://app.zenhub.com/workspaces/appointments-team-603fdef281af6500110a1691/issues/gh/department-of-veterans-affairs/va.gov-team/89322
https://app.zenhub.com/workspaces/appointments-team-603fdef281af6500110a1691/issues/gh/department-of-veterans-affairs/va.gov-team/89323
https://app.zenhub.com/workspaces/appointments-team-603fdef281af6500110a1691/issues/gh/department-of-veterans-affairs/va.gov-team/89329

**Comment by simiadebowale** [comment]
To help with writing tests, I need confirmation of the expected behavior in the following scenarios:
- **Only 1 Facility is returned in recent Locations sort**: ![Image](https://github.com/user-attachments/assets/4da8c2fd-dde8-40c0-ae4e-9d463d527f82)
- **No Facility returned in recent locations filter. (Patient has no recent appointments)**: Default back to the sort by home address option & remove the recent locations sort option![Image](https://github.com/user-attachments/assets/53154e0b-89e1-4ed0-a303-391cd6500fa3)
cc: @outerpress

**Comment by JunTaoLuo** [comment]
FYI I'm moving this back to blocked for now as we discussed last week.

**Comment by JunTaoLuo** [comment]
Carrying this over to sprint 2 for the remaining development work.

**Comment by simiadebowale** [comment]
@outerpress - this is ready for validation, you can test with judy.morrison@id.me in staging

**Comment by outerpress** [comment]
@simiadebowale I'm still testing but a couple of things:
- When testing with Judy, the note to update your home address doesn't show with the new recently visited option (see screen recording below). That should show for all options when the user has no home address
- Just to document this, the recently visited option should be at the top. As you said, this won't be possible until the feature is fully released and the Feature Flag is removed. Could you set up the Feature Flag removal ticket and add that note for testing?

https://github.com/user-attachments/assets/f7f30e0e-7a12-4985-bffa-62e22dfb5041

**Comment by outerpress** [comment]
@simiadebowale I think I see how this works, but want to confirm for testing:
- Look for appointments within 12 months of the current date (is that 12 months ahead and behind?)
- Facilities associated with appts furthest in the future will show first
- Facilities associated with appts furthest in the past will show last

Is that right?

**Comment by JunTaoLuo** [comment]
> [@simiadebowale](https://github.com/simiadebowale) I think I see how this works, but want to confirm for testing:
> 
> * Look for appointments within 12 months of the current date (is that 12 months ahead and behind?)
> * Facilities associated with appts furthest in the future will show first
> * Facilities associated with appts furthest in the past will show last
> 
> Is that right?

Since I worked on the backend vets-api changes for this, I'll add my replies here.

* We look for appointments from 1 year ago to 1 year in the future relative to the current day (so a total of 24 months worth of appointments)
* Correct, we simply sort by reverse chronological order 
* Correct, same as above

**Comment by outerpress** [comment]
Thanks @JunTaoLuo ! Two more clarifying questions: 
1. Does the lookup include facilities associated with telehealth appointments? Like if my furthest future appt is a telehealth appt scheduled through Daytona, would that show first in the list? (Fine if it does, just checking.)
2. Do community care appts have any impact on the lookup?

**Comment by outerpress** [comment]
Testing on Judy - looks almost perfect, so it may be human error on a Friday afternoon BUT - here's the first part of the list I'm seeing for Primary care:

<img width="516" alt="Image" src="https://github.com/user-attachments/assets/f8833589-a3e3-4cef-8d22-4d4823b84a72" />

Everything after this is alphabetized, so I'm focusing on these facilities. 

Upcoming appointments looks right - Cheyenne is the furthest out in the future, followed by Dayton, Fort Collins, Springfield.

Past appointments, there are a few Middletown, the oldest being August 2024.

The next is Sidney on Friday March 22, 2024, followed by Loveland, which is reverse the order shown in the list.

I clicked into all the video, ATLAS, and CC appts but couldn't find another Loveland. But it's definitely possible that I missed something. I'll test with some more users next week.

Oh, and all of the instances I could find of with Sidney or Loveland facilities are (conveniently) in the screenshot below:

<img width="586" alt="Image" src="https://github.com/user-attachments/assets/dee96ce6-dce1-4994-bdae-84c38420d7d2" />

**Comment by JunTaoLuo** [comment]
> Thanks [@JunTaoLuo](https://github.com/JunTaoLuo) ! Two more clarifying questions:
> 
> 1. Does the lookup include facilities associated with telehealth appointments? Like if my furthest future appt is a telehealth appt scheduled through Daytona, would that show first in the list? (Fine if it does, just checking.)
> 2. Do community care appts have any impact on the lookup?

1. We do not specify any limitations on which appointments to retrieve so we collect all facilities including telehealth ones.
2. Same as above, community care appointments are included as any other appointment.
3. I wanted to clarify one of my previous answers which is that if no appointments were found from 1 year in the past to 1 year in the future, we look at past appointments from 3 years ago to today.

**Comment by simiadebowale** [comment]
@outerpress @JunTaoLuo  - I think I figured out the issue, we're also [checking 1 year of pending appointment history](https://github.com/department-of-veterans-affairs/vets-api/blob/master/modules/vaos/app/services/vaos/v2/appointments_service.rb#L216). I checked and Judy has a pending appointment in October 2024. That explains why we're seeing Loveland before Sidney. The Sidney appointment is in March of 2024.
```
{
      "id": "401a084623bf239fba9e4d314d639488ffcf70b7c28676756cc0c87f33d23f8a",
      "type": "appointments",
      "attributes": {
        "id": "401a084623bf239fba9e4d314d639488ffcf70b7c28676756cc0c87f33d23f8a",
        "identifier": [
          {
            "system": "Appointment/",
            "value": "523938333137373939"
          },
          {
            "system": "http://www.va.gov/Terminology/VistADefinedTerms/409_85",
            "value": "983:17799"
          }
        ],
        "kind": "telehealth",
        "status": "proposed",
        "service_type": "ophthalmology",
        "service_types": [
          {
            "coding": [
              {
                "system": "http://veteran.apps.va.gov/terminologies/fhir/CodeSystem/vats-service-type",
                "code": "ophthalmology"
              }
            ]
          }
        ],
        "service_category": [
          {
            "coding": [
              {
                "system": "http://www.va.gov/Terminology/VistADefinedTerms/409_1",
                "code": "REGULAR",
                "display": "REGULAR"
              }
            ],
            "text": "REGULAR"
          }
        ],
        "reason_code": {
          "text": "station id: 983GD|preferred modality: VIDEO|phone number: 2011001000|email: test@id.me|preferred dates:10/01/2024 PM,10/04/2024 PM|reason code:OTHER_REASON|comments:This is a request test and cancel - LD "
        },
        "patient_icn": "1012845331V153043",
        "location_id": "983GD",
        "start": "2024-10-01T00:00:00Z",
        "minutes_duration": 0,
        "created": "2024-08-20T04:36:34Z",
        "requested_periods": [
          {
            "start": "2024-10-01T00:00:00Z",
            "end": "2024-10-01T00:00:00Z"
          }
        ],
        "cancellable": false,
        "extension": {
          "cc_location": {
            "address": {

            }
          }
        },
        "local_start_time": "2024-09-30T18:00:00.000-06:00",
        "station": "983",
        "ien": "17799",
        "contact": {
          "telecom": [
            {
              "type": "phone",
              "value": "2011001000"
            },
            {
              "type": "email",
              "value": "test@id.me"
            }
          ]
        },
        "patient_comments": "This is a request test and cancel - LD",
        "reason_for_appointment": "My reason isn’t listed",
        "preferred_dates": [
          "Tue, October 1, 2024 in the afternoon",
          "Fri, October 4, 2024 in the afternoon"
        ],
        "preferred_modality": "Video",
        "location": {
          "id": "983GD",
          "type": "appointments",
          "attributes": {
            "id": "983GD",
            "vista_site": "983",
            "vast_parent": "983",
            "type": "va_health_facility",
            "name": "Loveland VA Clinic",
            "classification": "Primary Care CBOC",
            "timezone": {
              "time_zone_id": "America/Denver"
            },
            "lat": 40.412647,
            "long": -105.003815,
            "website": "https://www.cheyenne.va.gov/locations/Loveland.asp",
            "phone": {
              "main": "970-962-4900"
            },
            "physical_address": {
              "type": "physical",
              "line": [
                "5200 Hahns Peak Drive"
              ],
              "city": "Loveland",
              "state": "CO",
              "postal_code": "80538-8852"
            },
            "health_service": [
              "Audiology",
              "EmergencyCare",
              "Gynecology",
              "MentalHealthCare",
              "Nutrition",
              "Ophthalmology",
              "Optometry",
              "Podiatry",
              "PrimaryCare",
              "SpecialtyCare",
              "WomensHealth"
            ],
            "ehr_system": "VistA"
          }
        },
        "type": "REQUEST",
        "modality": "vaVideoCareAtHome",
        "past": true,
        "future": false,
        "pending": true
      }
    }
```

**Comment by outerpress** [comment]
> we're also [checking 1 year of pending appointment history](https://github.com/department-of-veterans-affairs/vets-api/blob/master/modules/vaos/app/services/vaos/v2/appointments_service.rb#L216)

@simiadebowale  I was wondering if that was the case. I think that's okay. We're casting a pretty wide net, but it makes sense if someone requests an appointment at a facility, they will want/expect to schedule or request there again.

**Comment by outerpress** [comment]
@simiadebowale could you turn this on for the rest of the users on staging? Aside from the first issue I mentioned in [this comment](https://github.com/department-of-veterans-affairs/va.gov-team/issues/80213#issuecomment-2722965777), the only other thing I'm concerned about is load time. Looks like it's around 20 seconds for Judy, but we know she's an outlier. So I want to test on a few others.

**Comment by simiadebowale** [comment]
@outerpress - I took out that text but later realized that I didn't need to, I'll make a quick update to that. I have turned on the flag for all staging users

**Comment by outerpress** [comment]
Looking at a few more users, I'm running into an issue with sorting with Alfred Castillo. I don't see Cheyenne anywhere in his appointment list (upcoming or past), but it does show in the facility list above Daytona, which he does have appts at - see screenshot below.

Sorting looks good for:
* Claudia
* Ben Morgan
* Patient 997
* Patient 999

![Image](https://github.com/user-attachments/assets/8f25a006-212c-4631-be9d-1224734bd044)

**Comment by outerpress** [comment]
@simiadebowale notes from standup today:
* Likely the reason Cheyenne is showing is that he had a request at that facility that's no longer displayed in the list
* We can limit the lookback for pending appointments since those should be resolved quickly into appointments. Let's cut that back to 30 days of pending appts
* Once that's done I'll measure the load times for the location pages again. Here's what I have for the current state:

Judy - 20s
Claudia - 14s
Ben Morgan - 10s
Patient 997 - 7s
Patient 999 - 14s
Alfred Castillo - 8s

**Comment by outerpress** [comment]
@simiadebowale I retested the users with the feature off:

- Claudia - 1.2s
- Ben Morgan - 1.6s
- Patient 997 - 1.2s
- Patient 999 - 1.2s
- Alfred - 1.5s

Judy is still at 20s with the feature on.

**Comment by simiadebowale** [comment]
@outerpress - Noted, I'm putting together some notes but while testing, the recentLocations endpoint response time is coming up around 10s for Judy. 

![Image](https://github.com/user-attachments/assets/546206a3-0589-4213-bfa8-3f5bb4d8bc2c)

**Comment by simiadebowale** [comment]
We're collaborating with the backend team to implement caching for the fetch past appointments call. We'll revisit load times once this work is complete

**Comment by outerpress** [comment]
I'm taking my name off this one to clean up our board for end of sprint, please reassign to me when ready for another review. Thanks!

**Comment by outerpress** [comment]
@simiadebowale this one gets unblocked I have one more content change request:

Change `By recent locations` to `Recently visited` - that'll help clarify the sort criteria.

Would it be possible to fit that in? If not, no prob, we can add it into another ticket. Thanks!

**Comment by outerpress** [comment]
@simiadebowale sorry to pile on the comments. Along with the text changes, it looks like the events aren't firing that show changes to the dropdown - see screenshots below. Should I set up a separate ticket to spike on those/get them hooked up?

It's possible GA is doing something weird, but I only see them firing last year between May 6 and September 6...

<img width="1273" alt="Image" src="https://github.com/user-attachments/assets/7417ecff-4aba-4557-8372-d230019047cb" />

<img width="1273" alt="Image" src="https://github.com/user-attachments/assets/68e7f507-6a9f-469f-b573-0c533a788e71" />

**Comment by outerpress** [comment]
Oh, I think I just had the wrong events then - those are firing since March. I didn't realize we had new ones. I think `vaos-updated-sort--` will give us what we need, as long as we have a related event for recently visited when we release. Thank you!

**Comment by outerpress** [comment]
#106715 is done so moving back to validation

**Comment by outerpress** [comment]
@simiadebowale @JunTaoLuo I retested this since [#106715](https://github.com/department-of-veterans-affairs/va.gov-team/issues/106715) was completed and am still seeing some pretty slow page load times. Let's talk at standup next week to figure out next steps.

| User        | recentLocations call  | full page load  | 
| ----------- | --------------------- | --------------- |
| Judy        | `10.57s`              | `17.5s`         | 
| Claudia     | ` 9.22s`              | `12.0s`         |                                            
| Ben Morgan  | `10.84s`              | `13.5s`         |
| Patient 997 | ` 4.55s`              | ` 7.0s`         |
| Patient 999 | ` 8.39s`              | `12.0s`         |
| Alfred      | ` 6.09s`              | `13.5s`         |
| Brock       | ` 2.00s`              | ` 6.0s`         |

**Comment by outerpress** [comment]
@simiadebowale @JunTaoLuo  I dropped the results of the latest test below, along with the previous test to save us digging through comments. It looks like the load time went down a little for some, up a little for others, but is still much higher with the recent locations call (at least with this imperfect testing method).

**New test**

| User        | recentLocations call  | full page load w/ recentlocations | full page load w/o recent locations |
|-------------|-----------------------|-----------------------------------|-------------------------------------|
| Judy        |  `10.57s`             | `17.5s`                           | ` n/t`                              |
| Claudia     |  ` 9.22s`             | `12.0s`                           | ` >2s`                              |
| Ben Morgan  |  `10.84s`             | `13.5s`                           | `1.4s`                              |
| Patient 997 |  ` 4.55s`             | ` 7.0s`                           | `1.6s`                              |
| Patient 999 |  ` 8.39s`             | `12.0s`                           | `1.4s`                              |
| Alfred      |  ` 6.09s`             | `13.5s`                           | `3.5s`                              |
| Brock       |  ` 2.00s`             | ` 6.0s`                           | `2.2s`                              |   


**Previous test**
| User        | recentLocations call               | full page load w/ recent locations  | full page load w/o recent locations |
|-------------|------------------------------------|-------------------------------------|-------------------------------------|
| Judy        | `n/t`                              | `20s`                               | ` n/a`                              |
| Claudia     | `n/t`                              | `14s`                               | `1.2s`                              |
| Ben Morgan  | `n/t`                              | `10s`                               | `1.6s`                              |
| Patient 997 | `n/t`                              | ` 7s`                               | `1.2s`                              |
| Patient 999 | `n/t`                              | `14s`                               | `1.2s`                              |
| Alfred      | `n/t`                              | ` 8s`                               | `1.5s`                              |
| Brock       | `n/t`                              | `n/t`                               | `n/t`                               |

**Comment by JunTaoLuo** [comment]
Hmm interesting, I'll need to investigate a little further to figure out why there's still a huge discrepancy in load times outside of the `recentLocations` call by itself.

**Comment by simiadebowale** [comment]
I noticed we're making multiple calls to the facilities endpoint throughout the flow. We'll look into consolidating them to see if that helps improve load times.

**Comment by simiadebowale** [comment]
I have created a ticket for the enhancement described above: https://github.com/department-of-veterans-affairs/va.gov-team/issues/110235

**Comment by simiadebowale** [comment]
@outerpress - I've turned the flag on for all users in staging

**Comment by outerpress** [comment]
Numbers are below - looks like an improvement overall compared to the [last test](https://github.com/department-of-veterans-affairs/va.gov-team/issues/80213#issuecomment-2886948685) 👍 

The numbers are still high, but we can go ahead with a [phased release](https://github.com/department-of-veterans-affairs/va.gov-team/issues/111431). Let's talk about dates and tracking metrics next week.

| User        | recentLocations call  | full page load w/ recentlocations | full page load w/o recent locations |
|-------------|-----------------------|-----------------------------------|-------------------------------------|
| Judy        |  `10.33s`             | `16.00s`                           | ` n/t`                             |
| Claudia     |  ` 9.83s`             | `11.00s`                           | `1.8s`                             |
| Ben Morgan  |  ` 7.21s`             | ` 8.50s`                           | `1.6s`                             |
| Patient 997 |  ` 4.76s`             | ` 6.25s`                           | `1.2s`                             |
| Patient 999 |  ` 4.95s`             | ` 6.25s`                           | `1.1s`                             |
| Alfred      |  ` 7.00s`             | `10.50s`                           | `3.0s`                             |
| Brock       |  ` 2.29s`             | ` 4.00s`                           | `2.0s`                             |

@simiadebowale @JunTaoLuo
